src/components/ProductDetails/ProductDetails.js

```diff
@@ -0,0 +1,147 @@
+import React, { useState } from "react";
+import { useParams, useNavigate } from "react-router-dom";
+import { useProducts } from "../../contexts/useProducts";
+import "./productDetails.css";
+import axios from "axios";
+import { Toast } from "../Toast/Toast";
+import { AddToCartButton } from "../Products/AddToCartButton";
+
+export function ProductDetails() {
+  const { id } = useParams();
+  let navigate = useNavigate();
+  const [toast, setToast] = useState({
+    status: false,
+    heading: "",
+    msg: "",
+    error: false,
+  });
+
+  const { filteredData, state, dispatch } = useProducts();
+
+  const product = filteredData.find((item) => item._id === id);
+
+  async function handleCartAdd(product, state, setCartLoader) {
+    const isProduct = !!state.cartItems.find(
+      (item) => item.product._id === product._id
+    );
+    if (isProduct) {
```

Comment:
API requests can be made using separate util functions.
What I did was I created a separate server request file and all my axios call functions were present there.